### DISCUSSION TITLE
How do you read a blob as image from vercel storage?

### DISCUSSION THREAD

**Discussion by JeremiahDuane** [vercel/community#4060] [Help] [answered]
### Summary

I went through the [quickstart guide for blobs](https://vercel.com/docs/storage/vercel-blob/quickstart), and was able to successfully create a blob. I verified it exists by looking up the blob with in the vercel dashboard with the url generated by the example code. 

How do I now read/render a blob image in `next.js?` I thought I could simply insert the url in the `src` block of `next.js` image tags, but that didn't work (the alt text displayed, but no console errors). 

I also tried just following the link, but that just downloads a faulty jpeg file (previewer fails to open it). 

What am I doing wrong?

### Steps to Reproduce

1. Upload blob to vercel storage
2.  Read how?

**Comment by afrolisbon** [comment]
I have the same question, visiting the url only downloads the file in a new blank window, I wished this was covered in their documentation.

**Reply by JeremiahDuane** [reply]
The file itself is actually corrupted though, at least for me. My computer fails to open the downloaded files.

I'm beginning to think this is an encryption bug. I reached out to support, but haven't heard back yet.

**Reply by afrolisbon** [reply]
Mine is returning my base64 image file properly, if I copy paste the content from the downloaded file into my browser I do see the image rendered. If I need to an API call before using my image stored into that blob then I don't see the point of using it at all..

**Reply by JeremiahDuane** [reply]
You do??? I must be doing it wrong. That is great news. I'm pretty sure there's syntax for displaying a base64 image in HTML... just need to figure out what it is. I really wish this was in the documentation.

**Reply by afrolisbon** [reply]
It should already work from the get go with base64 images. 
On my test json DB I had images stored directly with that format and I could read them directly using <Image src /> without having to convert anything. 

To make the DB lighter, I switched to use the blob url returned from the API response and now I'm trying to read it on my frontend. 

I'm not sure if it's mean to be used this way or just meant to download the file. 

After all it's in Beta but yeah let's see..

**Reply by JeremiahDuane** [reply]
Alright, this is probably a dumb question, but how are you viewing it in browser? I try to open the file in Chrome and with Mac Preview and it keeps saying there is an issue with the file. 

When I open the file, its not even Base64... its Machine Code.

**Reply by afrolisbon** [reply]
Hi, I have to copy paste the content of the Document which is a base64 image into my Chrome URL bar. 
Are you sure your file is uploaded properly?

```
const handleImageChange = async (event: React.ChangeEvent<HTMLInputElement>) => {

const file = event.target.files?.[0];
const reader = new FileReader();

reader.onload = async () => {
 if (reader.result && typeof reader.result === 'string') {
  const imageDataUrl = reader.result as string;  //Your base64 image
  setImage(imageDataUrl); //Set state
 }
};
}
```

**Reply by JeremiahDuane** [reply]
I just used the vercel quickstart: https://vercel.com/docs/storage/vercel-blob/quickstart

They don't use a file reader though. Maybe I should know to use one, but I am no that smart lol. I will look up how to add a file reader to my input.

**Comment by afrolisbon** [comment]
After reading trough again the documentation for Vercel Blob,

>  Headers that enhance security by preventing unauthorized downloads, blocking external content from being embedded, and protecting against malicious file type manipulation, are enforced on each blob. They are:
> 
> Content-Disposition: attachment; filename="filename.extension"

This is why you can't access the image directly via the blob URL from the response. 

If we had the option to change the Content-Disposition to be **inline** then it would work as expected. 
I don't know if this is something that is work in progress. 

Source: https://vercel.com/docs/storage/vercel-blob#security

**Reply by JeremiahDuane** [reply]
I did it! Plus, I was reading the file incorrectly. After I started reading the file correctly, I was able to access the images correctly too. 

I saw that part in the documentation, but you can still work around the issue by fetching it again. 

You can do it like this in Next.js:

```
            // newBlob.url is the url associated with the image, the blob is the payload that is returned from the server after upload. 
            const imageResponse = await fetch(newBlob.url, {
              method: "GET",
            });
            const newImage = await imageResponse.text();
            setImage(newImage);
```

This parses the actual text from the file, instead of just downloading it. Set that as the SRC of your image tags!

**Reply by JeremiahDuane** [reply]
Also, if you want an easy pure JS example: 

Your JS:
```
const myURL = "https://eqlotw86esektkgo.public.blob.vercel-storage.com/llorkcir-5AJYcJtWz78QClo0WH0QUB4V1RJ6bQ.txt"

const getImage = async() => {
  const image = await fetch(myURL, {
    method: "GET"
  });
  const data = await image.text();
  document.getElementById("gotchu").src = data;
}
getImage()

```
Your HTML:
```
<h1>Attempt #2</h1>
<img id="gotchu" />
```

https://stackoverflow.com/questions/77145001/how-can-i-use-a-base64-encoded-image-in-a-text-file-as-an-html-image/77145089#77145089

**Reply by afrolisbon** [reply]
Yes nice one! 

The only thing is my goal was to avoid having to fetch the image before seeing it. 
It adds unnecessary API requests to the app especially if you have a list of 15 News Posts + to load on a page. 

Microsoft Blob Azure does offer the option to set the headers with contentDisposition: 'inline', which means you'd be able to directly use the url inside your image src without having to run it inside a function beforehand. 

Unfortunately Vercel doesn't seem to allow this for the moment. 

But I'm happy if this solution works for you!

**Reply by JeremiahDuane** [reply]
Gotcha, yeah, it's certainly not ideal. I already saw that part of the documentation though, so I wasn't expecting anything more at this point. I do hope they allow other content dispositions in the future, though.

**Reply by afrolisbon** [reply]
Yeah, for now I switched to Azure blob storage. Cheers.

**Comment by al222cv** [comment]
To get the Vercel Blob url to work with next/image, I had to create an api endpoint that takes in the Blob url, fetches the Blob, then return the Blob with the Content type image/* 


```
// app/api/image/route.tsx

import { NextRequest, NextResponse } from 'next/server'

export async function GET (request: NextRequest) {
  const url = request.nextUrl.searchParams.get('url')
  const res = await fetch(url!)
  const blob = await res.blob()
  const headers = new Headers()
  headers.set('Content-Type', 'image/*')
  return new NextResponse(blob, { status: 200, statusText: 'OK', headers })
}
```


Then use it like this: 

```
import Image from 'next/image'

<Image alt='Vercel Blob Image' src={`/api/image?url=${myVercelBlobImageUrl}`} fill />
```

This should not be necessary in my opinion...

**Comment by molebox** [comment, marked as answer]
I folks, we are updating the docs to reflect how to upload and view images stored in Vercel Blob using Server Actions. TLDR before we ship them, is that you can add the hostname to next.config. This will allow you to pull your images stored in blob, and use the url directly in Next.js Image component

```
/** @type {import('next').NextConfig} */
const nextConfig = {
  images: {
    remotePatterns: [{ hostname: "*.public.blob.vercel-storage.com" }],
  },
}

module.exports = nextConfig
```

**Reply by luvr-dev** [reply]
I'm little confused.  I manually upload images and storing the file names in my database.
I want to display them using a normal <Image ... with src
Assumed I'm going to see some "base url" somewhere and do something like <Image src={`${base_url}/{product_123_image.png`}

**Comment by codflexdez** [comment]
Hi all,
How would you download the image from the blob storage ?
